gemd/entity/link_by_uid.py

```diff
@@ -28,27 +29,46 @@ def __repr__(self):
         return str({"scope": self.scope, "id": self.id})
 
     @classmethod
-    def from_entity(cls, entity, name="auto"):
+    def from_entity(cls, entity, name=None, *, scope=None):
         """
-        Create LinkByUID from in-memory object using id with scope 'name'.
+        Create LinkByUID from in-memory object.
+
+        - If there exists an id with scope (default 'auto'), the LinkByUID object will be built
+          with that scope.
+        - If there is no id with scope, an arbitrary scope-id pair will be chosen.
+        - If the object has no uids, the object will be mutated to include a uid (UUID4) with the
+          chosen scope and the LinkByUID object will be built with that scope and id.
 
         Parameters
         ----------
         entity: BaseEntity
             The entity to substitute with a LinkByUID
-        name: str, optional
-            The scope of the id.
+        name: str, optional (Deprecated)
+            The desired scope of the id.
+        scope: str, optional
+            The desired scope of the id.
 
         Returns
         -------
         LinkByUID
             A link object that references `entity` through its scope and id.
 
         """
-        if name in entity.uids:
-            scope, uid = name, entity.uids[name]
+        if name is None:
+            if scope is None:
+                scope = "auto"  # set default
+        elif scope is None:
+            warn("The positional argument 'name' is deprecated.  When selecting a default scope, "
+                 "use the 'scope' keyword argument.", DeprecationWarning)
+            scope = name
+        else:
+            if scope is not None:
+                raise ValueError("Specify the 'name' parameter or 'scope' parameter, not both.")
```

Comment:
I don't actually think there's a logic error here, but this code took me an absurdly long time to parse for correctness. Could we emulate a pattern matching style with something like 
```suggestion
        if 
        if name is None and scope is None:               # don't break backwards compatibility when the user does nothing
            scope = "auto"
        elif name is not None and scope is None:      # deprecation warning existing users using only the "name" param
            warn("The positional argument 'name' is deprecated.  When selecting a default scope, "
                 "use the 'scope' keyword argument.", DeprecationWarning)
            scope = name
        elif name is not None and scope is not None:  
                raise ValueError("Specify the 'name' parameter or 'scope' parameter, not both.")
```

Reply:
Typo in the above, but code following this intent is now pushed.